DM-29708: Make ButlerHttpURI compatible with non-webDAV HTTP backends

bgounon opened the pull request:
This update allows the user to make simple requests against pure http backends without ButlerHttpURI complaining
+ reworked session creation a little to make authentication (via env variables) optional 
+ cleaned up old useless code from before the ButlerURI rework

bgounon committed:
Cleanup old unused methods

bgounon committed:
Unauthenticated and non-webdav backend support

bgounon committed:
Fix typo in docstring

timj commented on python/lsst/daf/butler/core/_butlerUri/http.py:
```diff
@@ -84,11 +84,22 @@ def getHttpSession() -> requests.Session:
+    try:
+        ca_bundle = os.environ['LSST_BUTLER_WEBDAV_CA_BUNDLE']
+    except KeyError:
+        log.warning("Environment variable LSST_BUTLER_WEBDAV_CA_BUNDLE is not set: "
```
I don't think this statement about https is true. I can get the content from a GitHub web page using ButlerURI just fine so this statement is not specifically about https.  I think this warning should only happen if you are connecting to a webdav backend (I think you have code to know) with `isWebdavEndpoint`

bgounon replied:
You are right, but it is not specifically about webdav neither. I changed the statement to make it clearer that it is mostly useful to trust non-default CAs, and made it a debug log instead of a warning log.

timj commented on python/lsst/daf/butler/core/_butlerUri/http.py:
```diff
@@ -84,11 +84,22 @@ def getHttpSession() -> requests.Session:
     except KeyError:
-        raise KeyError("Environment variable LSST_BUTLER_WEBDAV_AUTH is not set, "
-                       "please use values X509 or TOKEN")
+        log.debug("Environment variable LSST_BUTLER_WEBDAV_AUTH is not set, "
```
I'd only write this message if it's a webdav server.

bgounon replied:
My problem here is that getHttpSession() currently is not aware of the server URL, so we cannot internally check if the server is a webdav endpoint or not. I could add an optional parameter to getHttpSession() though, but even pure http servers may require x509/token authentication, so maybe it could be useful in some cases, even if the backend is not webdav.
And in that case maybe change the name of the env variable to something like LSST_BUTLER_HTTP_AUTH

timj replied:
I think the ability to add new CAs is worthwhile but we shouldn't complain if someone is doing https somewhere on GitHub and uses standard certificates. Now that it's a debug I'm not worried about the log message confusing people when it's not relevant.

timj commented on python/lsst/daf/butler/core/_butlerUri/http.py:
```diff
@@ -376,6 +287,10 @@ def size(self) -> int:
     def mkdir(self) -> None:
         """Create the directory resource if it does not already exist."""
+        # Only available on WebDAV backends
+        if not isWebdavEndpoint(self.scheme + "://" + self.netloc):
```
This scheme+netloc code seems like it needs to be a general method of ButlerURI that can be reused.

Since ButlerURI is now immutable, I think you probably want an `self.is_webdav_endpoint` `@property` that will test once on the server but then cache the response (it's never going to change).

timj reviewed (changes requested):
This looks okay but I'd like some minor changes:

* Add an is_webdav property that will cache the response from the server so you aren't continually asking the server if it's webdav each time (ButlerURI is immutable).
* Only warn about missing env vars if it's a webdav server.
* Consider adding a generic `ButlerURI.root()` method that returns the root URI (no path) [although if you've created the is_webdav property at least the current code will not be repeated in multiple places]

bgounon committed:
Add a generic ButlerURI.getrooturl() method

bgounon committed:
Add is_webdav_endpoint property to ButlerHttpURI

bgounon committed:
Clarify logs when a custom CA_BUNDLE is not set

bgounon commented:
@timj new version pushed, let me know if there are more changes to address

timj commented on python/lsst/daf/butler/core/_butlerUri/_butlerUri.py:
```diff
@@ -308,6 +308,16 @@ def geturl(self) -> str:
+        url : `str`
+            String form of root URI.
+        """
+        return self._uri.scheme + "://" + self._uri.netloc
```
To be consistent with the other APIs this should return a new ButlerURI and not a string.

timj commented on python/lsst/daf/butler/core/_butlerUri/http.py:
```diff
@@ -345,18 +263,16 @@ def session(self) -> requests.Session:
+    @property
+    def is_webdav_endpoint(self) -> bool:
+        """Check if the current endpoint implements WebDAV features"""
+        return isWebdavEndpoint(self.getrooturl())
```
Cache this in `self._is_webdav` so that we don't need the call to the server each time.
Something like:

```python
@property
def is_webdav_endpoint(self) -> bool:
    """..."""
    try:
        return self._is_webdav
    except AttributeError:
        pass
    self._is_webdav = isWebdavEndpoint(str(self.root_uri())
    return self._is_webdav
```

timj commented on python/lsst/daf/butler/core/_butlerUri/_butlerUri.py:
```diff
@@ -308,6 +308,16 @@ def geturl(self) -> str:
         """
         return self._uri.geturl()
 
+    def getrooturl(self) -> str:
```
I'd prefer this be `root_uri` I think.

timj commented on python/lsst/daf/butler/core/_butlerUri/http.py:
```diff
@@ -345,18 +263,16 @@ def session(self) -> requests.Session:
-            raise RuntimeError(f"Only Webdav endpoints are supported; got base URL '{baseURL}'.")
-
+        s = getHttpSession()
         ButlerHttpURI._session = s
```
Since we are storing this for all possible remote hosts we wouldn't be able to do the warnings per-host anyhow.

timj reviewed (approved):
Thanks. Minor comments only.

bgounon committed:
Return a ButlerURI with root_uri()

bgounon committed:
Cache is_webdav_endpoint value class-wide

timj commented on python/lsst/daf/butler/core/_butlerUri/http.py:
```diff
@@ -345,18 +263,21 @@ def session(self) -> requests.Session:
+            return ButlerHttpURI._is_webdav
+        except AttributeError:
+            pass
+        ButlerHttpURI._is_webdav = isWebdavEndpoint(self.root_uri())
```
Doesn't this mean that every subsequent ButlerURI thinks it's webdav even if it's for a completely different server? I think the cache should be indexed by the root_uri shouldn't it?

timj replied:
It might be easier to have a non-class function that uses an `functools.lru_cache` that takes the `self.root_uri()` as argument. You'll get caching for free (ButlerURI can now be used as a key).

bgounon replied:
I made `_is_webdav` class-wide to be coherent with what I do with `_session`, although I admit I'm unsure this is the best way to do it. In my tests, without using the class scope, `_session` and `_is_webdav` would be refreshed every time we change directory/file, even though it is the same endpoint/server.
I will look into `functools.lru_cache` which I didn't know before.
I'll also fix mypy/lint issues in an upcoming commit.

timj replied:
Global cache is absolutely fine but as written it was caching a single value regardless of the URI root. We need a per-root cache which can doesn't need to be attached to the class itself (hence lru_cache decorator on a standalone function).

timj commented:
@bgounon would you like me to take over this pull request? I think you are most of the way there and I just need to add the global cache by host name.

bgounon commented:
> @bgounon would you like me to take over this pull request? I think you are most of the way there and I just need to add the global cache by host name.

Hi @timj, sorry for letting this PR hanging like that. I've mostly been busy with Qserv and RSP matters recently and haven't taken the time to finalize things here. I have no problem with you taking over for the cache implementation, of course.
Otherwise, I expect I should be able to come back to it the week after next.

timj committed:
Remove trailing whitespace

timj committed:
Cache webdav endpoint check per host.

timj committed:
Stop worrying about Location in http URI

The code path was not being used so simplify things by removing
it.

timj committed:
Fix space in message

timj reviewed (approved):
@bgounon I've added the webdav caching so I think I'm happy to merge it now. I'll merge next week unless you say something. Thanks for the cleanup.

timj committed:
Add news fragment

timj committed:
Fix name of news fragment file

timj committed:
Move doc build to different github action

This should speed up the checks a little since we do not need
to wait to build the docs.

timj committed:
Ensure that changes are rendered into documentation
